### PR TITLE
fix: minimize the precision loss when dumping double to string

### LINKED ISSUE #711: Dumping double value to string can cause a precision loss
### Crow version:
v1.0+5

### How to reproduce
In this [test case](https://github.com/CrowCpp/Crow/blob/master/tests/unittest.cpp#L1041), change this value from `4.2` to `0.036303908355795146`, build and run this test again.
Log:
```
Expanded: `"0.036304" == "0.036303908355795146"`
```

### PR DESCRIPTION
**Pull request by junekhan**
This PR fixes #711

### CODE REVIEW

File: include/crow/json.h [hunk 1]
```diff
@@ -1851,10 +1851,11 @@ namespace crow
                             } f_state;
                             char outbuf[128];
 #ifdef _MSC_VER
-                            sprintf_s(outbuf, sizeof(outbuf), "%f", v.num.d);
+#define MSC_COMPATIBLE_SPRINTF(BUFFER_PTR, FORMAT_PTR, VALUE) sprintf_s((BUFFER_PTR), 128, (FORMAT_PTR), DBL_DECIMAL_DIG, (VALUE))
```
Comment (gittiver) [inline include/crow/json.h:1854]: I would not reintroduce the macro here

File: include/crow/json.h [hunk 2]
```diff
@@ -1851,10 +1851,11 @@ namespace crow
...
 #else
-                            snprintf(outbuf, sizeof(outbuf), "%f", v.num.d);
+#define MSC_COMPATIBLE_SPRINTF(BUFFER_PTR, FORMAT_PTR, VALUE) sprintf((BUFFER_PTR), (FORMAT_PTR), DECIMAL_DIG, (VALUE))
```
Comment (gittiver) [inline include/crow/json.h:1856]: I would not reintroduce the macro here

File: include/crow/json.h (hunk 1 above)
Comment (junekhan) [reply include/crow/json.h:1854]: @gittiver Do you have any suggestion?

Comment (gittiver) [reply include/crow/json.h:1854]: leave the sprintf_s and sprintfs together with the new format string and don't introduce the macro for this one time use.

File: include/crow/json.h (hunk 2 above)
Comment (gittiver) [reply include/crow/json.h:1856]: see first comment.

File: include/crow/json.h (hunk 1 above)
Comment (junekhan) [reply include/crow/json.h:1854]: Alright, there you go!

File: include/crow/json.h [hunk 3]
```diff
@@ -1851,9 +1851,9 @@ namespace crow
                             } f_state;
                             char outbuf[128];
 #ifdef _MSC_VER
-                            sprintf_s(outbuf, sizeof(outbuf), "%f", v.num.d);
+                            sprintf_s(outbuf, 128, "%.*g", DBL_DECIMAL_DIG, v.num.d);
```
Comment (gittiver) [inline include/crow/json.h:1854]: sprintf_s(outbuf, sizeof(outbuf), "%.*g", DBL_DECIMAL_DIG, v.num.d);

File: include/crow/json.h [hunk 4]
```diff
@@ -1851,9 +1851,9 @@ namespace crow
...
 #else
-                            snprintf(outbuf, sizeof(outbuf), "%f", v.num.d);
+                            sprintf(outbuf, "%.*g", DECIMAL_DIG, v.num.d);
```
Comment (gittiver) [inline include/crow/json.h:1856]: snprintf(outbuf, sizeof(outbuf), "%.*g",DECIMAL_DIG, v.num.d);

**Review by gittiver — CHANGES_REQUESTED** [review]
see comments

File: include/crow/json.h [hunk 5]
```diff
@@ -1851,9 +1851,9 @@ namespace crow
                             } f_state;
                             char outbuf[128];
 #ifdef _MSC_VER
-                            sprintf_s(outbuf, sizeof(outbuf), "%f", v.num.d);
+                            sprintf_s(outbuf, sizeof(outbuf), "%.*g", DBL_DECIMAL_DIG, v.num.d);
 #else
-                            snprintf(outbuf, sizeof(outbuf), "%f", v.num.d);
+                            snprintf(outbuf, sizeof(outbuf), "%.*g", DECIMAL_DIG, v.num.d);
```
Comment (gittiver) [inline include/crow/json.h:1856]: shouldn't DBL_DECIMAL_DIG also used here?

Comment (junekhan) [reply include/crow/json.h:1856]: Interesting.
`DECIMAL_DIG` is under C++11 standard, while `DBL_DECIMAL_DIG` is C++17 and it doesn't work in my building environment. Let me make the change and figure out my issue later on.

File: include/crow/json.h [hunk 6]
```diff
@@ -1851,9 +1851,9 @@ namespace crow
                             } f_state;
                             char outbuf[128];
 #ifdef _MSC_VER
-                            sprintf_s(outbuf, sizeof(outbuf), "%f", v.num.d);
+                            sprintf_s(outbuf, sizeof(outbuf), "%.*g", DBL_DECIMAL_DIG, v.num.d);
 #else
-                            snprintf(outbuf, sizeof(outbuf), "%f", v.num.d);
+                            snprintf(outbuf, sizeof(outbuf), "%.*g", DBL_DECIMAL_DIG, v.num.d);
```
Comment (gittiver) [inline include/crow/json.h:1856]: As we have set C++ 11, build fails with it, you are right. means we have to use DECIMAL_DIG here.

**Comment by gittiver** [comment]
a unit test on the template tests fails - it expect exact values not something with epsilon. 
Needs to be adressed before merging the PR.

**Review by gittiver — CHANGES_REQUESTED** [review]
Unittest template_tests fails.

**Review by gittiver — CHANGES_REQUESTED** [review]
unit test template test fails

**Comment by junekhan** [comment]
> a unit test on the template tests fails - it expect exact values not something with epsilon. Needs to be adressed before merging the PR.

Yes, I've made a patch for that, and the scope becomes much bigger. Please have a careful check to help me ensure the changes. BTW, some of the unit tests can run properly on my machine, so I have to resort to this github action. Thanks.

**Review by gittiver — APPROVED** [review]

**Pull request merged**